doc/build-overview.md

```diff
@@ -0,0 +1,251 @@
+* If the `--compiler` CLI arg is set, or the `compiler` config value
+  is set (and `--resolver` CLI arg is not set), ignore the wanted
+  compiler from the snapshot and use the specified wanted compiler
+* Parse `extra-deps` into a `Map PackageName PackageLocation`,
+  containing both mutable and immutable package locations. Parse
+  `packages` into a `Map PackageName ProjectPackage`.
+* Ensure there are no duplicates between these two sets of packages
+* Delete any packages from the snapshot packages that appear in
+  `packages` or `extra-deps`
+* Perform a left biased union between the immutable `extra-deps`
+  values and the snapshot packages. Ignore any settings in the
+  snapshot packages that have been replaced.
+* Apply the `flags` and `ghc-options` by name to these packages. If
+  any values are specified but no matching package is found, it's an
+  error.
+* We are now left with the following:
+    * A wanted compiler version
+    * A map from package name to immutable packages with package config (flags, GHC options, hidden)
+    * A map from package name to mutable packages as dependencies with package config
+    * A map from package name to mutable packages as project packages with package config
+
+## Get actual compiler
+
+Use the wanted compiler and various other Stack config values (not all
+listed here) to find the actual compiler, potentially installing it in
+the process.
+
+## Global package sources
+
+With the actual compiler discovered, list out the packages available
+in its database and create a map from package name to
+version/GhcPkgId. Remove any packages from this map which are present
+in one of the other three maps mentioned above.
+
+## Resolve targets
+
+Take the CLI args for targets as raw text values and turn them into
+actual targets.
+
+* Do a basic parse of the values into one of the following:
+    * Package name
+    * Package identifier
+    * Package name + component
+    * Directory
+* An empty target list is equivalent to listing the package names of
+  all project packages
+* For any directories specified, find all project packages in that
+  directory or subdirectories therefore and convert to those package
+  names
+* For all package identifiers, ensure that either the package name
+  does not exist in any of the three parsed maps from the "wanted
+  compiler" step above, or that the package is present as an immutable
+  dependency from Hackage. If so, create an immutable dependency entry
+  with default flags, GHC options, and hidden status, and add this
+  package to the set of immutable package dependencies.
+* For all package names, ensure the package is in one of the four maps
+  we have, and if so add to either the dependency or project package
+  target set.
+* For all package name + component, ensure that the package is a
+  project package, and add that package + component to the set of
+  project targets.
+* Ensure that no target has been specified multiple times.
```

Comment:
I think we will need an extra consistency step for internal libraries. Sometimes stack needs to use the mangled name (`z-package-internallibname-z..`), sometimes the `package:internallibname` one. But I think this will become obvious when doing the code changes.

Reply:
I'm going to copy your comment in as a FIXME.